src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,201 @@
+                    sparkContext);
+                _jvmObject = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                    "org.apache.spark.api.python.PythonRDD",
+                    "readBroadcastFromFile",
+                    javaSparkContext,
+                    path);
+            }
+            else
+            {
+                SparkConf sparkConf = sc.GetConf();
+                sc.EncryptionEnabled = bool.Parse(
+                    sparkConf.Get("spark.io.encryption.enabled",
+                    "false"));
+
+                var pythonBroadcast = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                    "org.apache.spark.api.python.PythonRDD",
+                    "setupBroadcast",
+                    path);
+
+                if (sc.EncryptionEnabled)
+                {
+                    throw new NotImplementedException(
+                            "Broadcast encryption is not supported yet.");
+                }
+                else
+                {
+                    WriteBroadcastValueToFile(sc, path, value);
+                }
+
+                _jvmObject = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
+                    "org.apache.spark.sql.api.dotnet.SQLUtils",
+                    "createBroadcast",
+                    sparkContext,
+                    pythonBroadcast);
+            }
+            _bid = (long)_jvmObject.Invoke("id");
+            JvmBroadcastRegistry.s_jvmBroadcastVariables.Add(_jvmObject);
+        }
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Get the broadcasted value.
+        /// </summary>
+        /// <returns>The broadcasted value</returns>
+        public T Value()
+        {
+            return (T)BroadcastRegistry.s_registry[_bid];
+        }
+
+        /// <summary>
+        /// Asynchronously delete cached copies of this broadcast on the executors.
+        /// If the broadcast is used after this is called, it will need to be re-sent to each
+        /// executor.
+        /// </summary>
+        public void Unpersist()
+        {
+            _jvmObject.Invoke("unpersist");
+        }
+
+        /// <summary>
+        /// Delete cached copies of this broadcast on the executors. If the broadcast is used after
+        /// this is called, it will need to be re-sent to each executor.
+        /// </summary>
+        /// <param name="blocking">Whether to block until unpersisting has completed</param>
+        public void Unpersist(bool blocking)
+        {
+            _jvmObject.Invoke("unpersist", blocking);
+        }
+
+        /// <summary>
+        /// Destroy all data and metadata related to this broadcast variable. Use this with
+        /// caution; once a broadcast variable has been destroyed, it cannot be used again.
+        /// This method blocks until destroy has completed.
+        /// </summary>
+        public void Destroy()
```

Comment:
do we have tests for this?

Reply:
I am working on it. I am unable to catch the Exception with `Assert.Throws<Exception>` after the broadcast variable is destroyed, I get the following error: System.Runtime.Serialization.SerializationException : Type 'Microsoft.Spark.E2ETest.UdfTests.BroadcastTests' in Assembly 'Microsoft.Spark.E2ETest, Version=0.9.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' is not marked as serializable.

In the complete stacktrace while debugging I do see the Exception from Spark something along the lines of: `Broadcast called after being destroyed..`